### ISSUE TITLE
Ideas for deeper git integration

### ISSUE THREAD

**Issue by HadrienG2** [spack/spack#8785]
Most recent C++ software is developed using some kind of Git forge (typically GitHub or a self-hosted service), and uses CMake as its low-level build system. Within their git repository, software releases are typically tracked using one of a few different tagging conventions, e.g. "v1.2.3".

This de facto standardization could in principle be leveraged by Spack in order to ease the job of packagers, in the following way: when a packager types in something like `spack create https://gitlab.cern.ch/acts/acts-core.git`, Spack could...

1. Auto-detect that the URL refers to a git repository (these typically end with ".git")
2. Make a local clone of said repository.
3. Enumerate the list of tags.
4. Filter those which use a typical release tagging convention (e.g. "X.Y.Z" or "vX.Y.Z")
5. Automatically generate `version(<version number>, commit=<hash>)` entries from release tags, plus one entry for the master branch.
6. Figure out that the package uses CMake (e.g. by detecting a CMakeLists file in the master branch)
7. Use CMake to infer project dependencies and configuration options (prior art: #3626).
8. Provide the packager in the end with a package.py that has...
    * The right name (no ".git" or "-git" remaining from the URL)
    * The right template (CMakePackage)
    * A pre-filled list of versions from the filtered git tags
    * A pre-filled list of dependencies from CMake FindPackages
    * A pre-filled list of variants from CMake's options.
    * A couple of FIXMEs for things which cannot be autogenerated.

In this way, a packager would only need to...

- Review the job of the automatic algorithm
- Add dependencies which have not been properly declared in the CMakeLists
- Give variants a nicer name/description which follows Spack conventions, and remove those which are not deemed interesting enough.
- Account for package-specific build idiosyncraties which are only documented in human language

A possible future extension of this design would be to add something like a `spack update acts-core` command, which would automatically re-analyse the project's git repository and adds new tags as new software releases come in.

What do you think about this general idea? I think reaching this level of automation would be desirable, because it would reduce the amount of mechanical and repetitive tasks which humans need to do, and increase the amount of time available for working on the hard packaging problems which only humans can do (e.g. figuring out missing dependencies). On the other hand, one negative side is that this adds more automagic to Spack, which will cost some development time to maintain and reduce in weirder failures when it gets something wrong.

**Comment by hartzell** [comment]
While I like automation as much as the next programmer, in instances like this it gives me pause.

I have a [possibly naive or un-founded] belief that the versions of a package have been built by and are known to work for at least someone, somewhere.   As things stand now, I (and perhaps others) feel like I should exercise a package with a new release before I PR it (not that I'm *perfect*, but...).

If/when we hoover up all of the tags in a repo, will people actually build them and take them out for a spin?

**Comment by goxberry** [comment]
I'd like to add a 9th item: add a Ninja backend for CMake, which would speed up package builds.

**Comment by HadrienG2** [comment]
@hartzell Building all available tags should be easy to automate. You are right that testing is, by nature, harder to automate, though. Adding it to the list of things which packagers would still need to do...

Which reminds me: does Spack have some support for automated package testing? I have not seen anything of this sort while skimming through the documentation, but perhaps I have missed it?

@goxberry I think this part could be a separate issue, but it is true that Ninja support would be nice!

**Comment by adamjstewart** [comment]
> does Spack have some support for automated package testing?

@HadrienG2 This is a long-standing feature request that has been in the works for a while now. I expect this to be finished before the 1.0 release. See https://github.com/spack/spack/projects/3

> add a Ninja backend for CMake, which would speed up package builds.

@goxberry CMake already has a Ninja backend, see https://github.com/spack/spack/blob/develop/lib/spack/spack/build_systems/cmake.py#L76,L84. See `archer` and `llvm-openmp-ompt` for packages that already make use of it.

**Comment by hartzell** [comment]
> [...] You are right that testing is, by nature, harder to automate, though. [...]

One of the wonderfully annoying things about submitting a package recipe to Homebrew is that they *must* have tests and they're insistent that those tests do more than just check that e.g. `app --version` works (believe me, I've tried....).  Here's a bit of the test I started putting together for `tidy-markdown` before they decided it would never get merged  (users should manage it via [sigh] `npm`).  If I had gone forward with it I would have had to beef up the assert statements.

```ruby
  test do
    (testpath/"test.md").write <<-EOS.undent
      #    Now is the time
      header1 | header 2 | long header number 3
      ---|---|---
      foo|bar mitzvah| baz
      eenie|meenie|myne-mo
    EOS
    cmd = ["cat", testpath/"test.md", "|", bin/"tidy-markdown"].join(' ')
    s = pipe_output(cmd)
    # check if it cleaned up the extra whitespace after the h1
    assert_match /# Now is the time/, s
  end
end
```

Tests like this in Spack would be annoyingly wonderful too!

**Comment by adamjstewart** [comment]
> One of the wonderfully annoying things about submitting a package recipe to Homebrew is that they must have tests

That explains why they have so many tests. Hmm... that does sound really nice. Don't give me any ideas 😛 

I think Spack is still too young for this, but maybe someday when it matures. For now we'll just have to rely on `make check` and `make installcheck`.

**Comment by goxberry** [comment]
@adamjstewart Yes, and the feature request is to make it possible for any CMake package to use the Ninja backend, not just the two you mentioned.

**Comment by adamjstewart** [comment]
@goxberry So what you want is to be able to tell Spack which generator to use for _all_ CMake packages, not on a per-package basis. I see. That should be doable.

**Comment by HadrienG2** [comment]
@goxberry @adamjstewart Beware that CMake's Ninja backend does not have exact feature parity with the GNU Make backend: there are CMake packages out there which will only build with make.

Therefore, what would be needed is more exactly a way to make Ninja the default, but force use of the GNU Make backend on a per-package basis.

**Comment by HadrienG2** [comment]
I don't have access to my collection of build recipes right now, let me give you some examples when I'm back at the lab. There was the fortran thing, but also other things like downloads which wouldn't start in ninja builds. (There is a worrying growing tendency for CMake projects to download their dependencies and build them instead of looking for them on the host).

**Comment by HadrienG2** [comment]
Took me longer than I thought, but I found the time to investigate my build script archives. In the end, I found two CMake + Ninja failure modes:

- Projects including Fortran code, which as you mention have a Kitware-provided workaround.
- ExternalProject requiring extra parameters when ninja is involved (see [this e-mail thread](https://cmake.org/pipermail/cmake/2015-April/060233.html)).

**Comment by goxberry** [comment]
@HadrienG2 re: the second point, the e-mail thread you linked to also suggests fixes for buggy `ExternalProject`-vendored dependencies.

In my limited experience, the combination of Ninja and ccache dramatically reduces repeat build times -- rebuilds that would take GNU Make minutes instead take Ninja + ccache seconds. And a big chunk of that decrease is Ninja, which also accelerates initial builds.

This combination is disruptively helpful. The biggest bottleneck to debugging a package-manager-provided package is repeatedly building the package from source. Lightning-fast rebuilds are the difference between debugging a spack package in a week versus an hour.

I realize that ccache is a big lift, and already addressed by other issues and PRs in Spack, which is why I put "make it possible to build any CMake package with Ninja" at the top of my "CMake/spack integration feature wish list" -- it would increase my throughput on new `CMakePackage` PRs by 5-10x.

**Comment by adamjstewart** [comment]
I think I mentioned this in another thread somewhere, but once https://github.com/spack/spack/projects/3 is complete, I would wholeheartedly support a PR that switches the default from "GNU Makefiles" to "Ninja" for all CMake packages. That project will allow us to build all CMakePackages before and after and see which ones break. We could then override the default in those packages if necessary.
